File: drivers/SmarActMCS2/SDK-2.1.3/SDK/C/examples/9_MCS2Example_MagneticDriver/MCS2Example_MagneticDriver.c
/*
* SmarAct MCS2 programming example: MagneticDriver
*
* This programming example demonstrates how to control
* electromagnetic driven positioners with the MCS2
*
* For a full command reference see the MCS2 Programmers Guide.
*
*/

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <inttypes.h>
#include "SmarActControl.h"

#if defined(_WIN32)
#define strtok_r strtok_s
#endif

SA_CTL_DeviceHandle_t dHandle;

void exitOnError(SA_CTL_Result_t result) {
    if (result != SA_CTL_ERROR_NONE) {
        SA_CTL_Close(dHandle);
        // Passing an error code to "SA_CTL_GetResultInfo" returns a human readable string
        // specifying the error.
        printf("MCS2 %s (error: 0x%04x).\nPress return to exit.\n", SA_CTL_GetResultInfo(result), result);
        getchar();
        exit(1);
    }
}

void waitWhile(int8_t channel, uint32_t mask) {
    uint32_t state;
    while (1) {
        SA_CTL_Result_t result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_STATE, (int32_t*)&state, 0);
        exitOnError(result);
        if ((state & mask) == 0) break;
    }
}

int main() {

    SA_CTL_Result_t result;

    printf("***************************************************************\n");
    printf("*  SmarAct MCS2 Programming Example (Electromagnetic Driver)  *\n");
    printf("***************************************************************\n");

    // Read the version of the library
    // Note: this is the only function that does not require the library to be initialized.
    const char *version = SA_CTL_GetFullVersionString();
    printf("SmarActCTL library version: %s.\n", version);

    // Find available MCS2 devices
    char deviceList[1024];
    size_t ioDeviceListLen = sizeof(deviceList);
    result = SA_CTL_FindDevices("", deviceList, &ioDeviceListLen);
    if (result != SA_CTL_ERROR_NONE) {
        printf("MCS2 failed to find devices.\n");
    }
    exitOnError(result);
    if (strlen(deviceList) == 0) {
        printf("MCS2 no devices found. Exit.\n");
        getchar();
        exit(1);
    }
    else {
        printf("MCS2 available devices:\n%s\n", deviceList);
    }
    // Open the first MCS2 device from the list
    char *ptr;
    strtok_r(deviceList, "\n", &ptr);
    char *locator = deviceList;
    result = SA_CTL_Open(&dHandle, locator, "");
    if (result != SA_CTL_ERROR_NONE) {
        printf("MCS2 failed to open \"%s\".\n", locator);
    }
    exitOnError(result);
    printf("MCS2 opened \"%s\".\n", locator);

    int8_t channel = 0;
    int32_t type, state, error;

    // Verify that there is an electromagnetic driven positioner available.
    result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_TYPE, &type, 0);
    exitOnError(result);
    result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_STATE, &state, 0);
    exitOnError(result);
    if (type != SA_CTL_MAGNETIC_DRIVER) {
        printf("MCS2 channel %d is not an electromagnetic driver channel, abort.\n", channel);
        SA_CTL_Close(dHandle);
        getchar();
        exit(1);
    } else if ((state & SA_CTL_CH_STATE_BIT_SENSOR_PRESENT) == 0) {
        printf("MCS2 no positioner connected to channel %d, abort.\n", channel);
        SA_CTL_Close(dHandle);
        getchar();
        exit(1);
    }

    // The amplifier of electromagnetic driver channels is disabled at startup and must be explicitly enabled
    // before being able to perform closed-loop movements. This also implicitly starts the phasing sequence of
    // the positioner if it is not already phased.
    // The SA_CTL_CH_STATE_BIT_AMPLIFIER_ENABLED channel state bit reflects the state of the amplifier.
    if ((state & SA_CTL_CH_STATE_BIT_AMPLIFIER_ENABLED) == 0) {
        printf("*******************************************************\n");
        printf("-> Press return to enable the amplifier (and start the phasing sequence).\n");
        getchar();
        result = SA_CTL_SetProperty_i32(dHandle, channel, SA_CTL_PKEY_AMPLIFIER_ENABLED, SA_CTL_TRUE);
        exitOnError(result);
        if ((state & SA_CTL_CH_STATE_BIT_IS_PHASED) == 0) {
            printf("MCS2 phasing...\n");
            // Note that no external force or displacement must be applied to the positioner while the sequence is running.
            // The phasing takes some time to complete. During this time the SA_CTL_CH_STATE_BIT_ACTIVELY_MOVING
            // channel state bit is set.
            // The channel generates a SA_CTL_EVENT_PHASING_FINISHED event once the sequence has finished.
            // In this example we poll the channel state to determine the end of the phasing sequence instead of
            // listening to events.
            waitWhile(channel, SA_CTL_CH_STATE_BIT_ACTIVELY_MOVING);
            // The SA_CTL_CH_STATE_BIT_IS_PHASED channel state flag should then be checked to verify that the
            // phasing sequence finished successfully.
            result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_STATE, &state, 0);
            exitOnError(result);
            if ((state & SA_CTL_CH_STATE_BIT_IS_PHASED) == 0) {
                printf("MCS2 could not establish the phasing reference for channel %d, abort.\n", channel);
                SA_CTL_Close(dHandle);
                getchar();
                exit(1);
            } else {
                printf("MCS2 phase reference found.\n");
            }
        }
    }
    // Perform referencing sequence if physical position is not known
    if ((state & SA_CTL_CH_STATE_BIT_IS_REFERENCED) == 0) {
        // Since the physical scale is not known yet no range limits can be used here.
        // Make sure to use a moderate move velocity for the referencing sequence.
        // Disable software range limits while referencing.
        result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_RANGE_LIMIT_MIN, 0);
        exitOnError(result);
        result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_RANGE_LIMIT_MAX, 0);
        exitOnError(result);
        result = SA_CTL_SetProperty_i32(dHandle, channel, SA_CTL_PKEY_REFERENCING_OPTIONS, 0);
        exitOnError(result);
        // The move velocity and acceleration properties also define the parameters for the referencing.
        result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_MOVE_VELOCITY, 2000000000);
        exitOnError(result);
        result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_MOVE_ACCELERATION, 100000000000);
        exitOnError(result);
        printf("*******************************************************\n");
        printf("-> Press return to start the referencing.\n");
        getchar();
        printf("MCS2 referencing...\n");
        result = SA_CTL_Reference(dHandle, channel, 0);
        exitOnError(result);
        waitWhile(channel, SA_CTL_CH_STATE_BIT_REFERENCING);
        // Check if the referencing was successful.
        result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_STATE, &state, 0);
        exitOnError(result);
        // The SA_CTL_CH_STATE_BIT_MOVEMENT_FAILED indicates a failed referencing.
        if ((state & SA_CTL_CH_STATE_BIT_MOVEMENT_FAILED) != 0) {
            // The channel error property may then be read to determine the reason of the error.
            result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_ERROR, &error, 0);
            exitOnError(result);
            printf("MCS2 referencing failed: %s (error: 0x%04x), abort.\n", SA_CTL_GetResultInfo(error), error);
            SA_CTL_Close(dHandle);
            getchar();
            exit(1);
        } else {
            printf("MCS2 reference found.\n");
        }
    }
    // Note that electromagnetic driven positioners can reach very high velocities!
    // Moving into a physical endstop with high velocity can damage the positioner.
    // Therefore it is recommended to use the software range limits to define the active movement
    // range of the positioner. When commanding the positioner towards a limit the positioner is
    // decelerated to zero velocity in a way that it comes to a halt on the specified limit position.
    // Note: adjust this limits to match the movement range of your positioner!
    result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_RANGE_LIMIT_MIN, -20000000000);
    exitOnError(result);
    result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_RANGE_LIMIT_MAX, 20000000000);
    exitOnError(result);

    printf("*******************************************************\n");
    printf("-> Press return to start the movement.\n");
    getchar();
    // Set move mode to absolute movement.
    // Note: Open-loop movement is not available for electromagnetic driver channels.
    // (only the SA_CTL_MOVE_MODE_CL_ABSOLUTE and SA_CTL_MOVE_MODE_CL_RELATIVE move modes are valid)
    result = SA_CTL_SetProperty_i32(dHandle, channel, SA_CTL_PKEY_MOVE_MODE, SA_CTL_MOVE_MODE_CL_ABSOLUTE);
    exitOnError(result);
    // Set move velocity and acceleration.
    // Note that the velocity and acceleration control must be used for all movements. A velocity / acceleration
    // value of 0 (to disable the velocity / acceleration control) is invalid for magnetic driver channels.
    result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_MOVE_VELOCITY, 10000000000);
    exitOnError(result);
    result = SA_CTL_SetProperty_i64(dHandle, channel, SA_CTL_PKEY_MOVE_ACCELERATION, 10000000000);
    exitOnError(result);

    int64_t positions[] = {2000000000,-2000000000,10000000000,-10000000000}; // in pm
    for (uint32_t i = 0; i < 4; i++) {
        printf("MCS2 move channel %d to position: %" PRId64"...\n", channel, positions[i]);
        result = SA_CTL_Move(dHandle, channel, positions[i], 0);
        exitOnError(result);
        waitWhile(channel, SA_CTL_CH_STATE_BIT_ACTIVELY_MOVING);
        // Check if the movement was successful.
        result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_STATE, &state, 0);
        exitOnError(result);
        // The SA_CTL_CH_STATE_BIT_MOVEMENT_FAILED channel state flag indicates a failed movement.
        if ((state & SA_CTL_CH_STATE_BIT_MOVEMENT_FAILED) != 0) {
            // The channel error property may then be read to determine the reason of the error.
            result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_CHANNEL_ERROR, &error, 0);
            exitOnError(result);
            printf("MCS2 movement failed: %s (error: 0x%04x).\n", SA_CTL_GetResultInfo(error), error);
            break;
            // Alternatively, the following channel state flags may be tested to determine the reason
            // of a failed movement:
            // - SA_CTL_CH_STATE_BIT_END_STOP_REACHED
            // - SA_CTL_CH_STATE_BIT_RANGE_LIMIT_REACHED
            // - SA_CTL_CH_STATE_BIT_FOLLOWING_LIMIT_REACHED
            // - SA_CTL_CH_STATE_BIT_POSITIONER_OVERLOAD
            // - SA_CTL_CH_STATE_BIT_POSITIONER_FAULT
            // See the MCS2 Programmers Guide for more information on the specific channel state flags.
        }
        // The electromagnetic driver monitors the output current of each channel to detect an overload condition
        // of the positioner. This prevents thermal overheating and potential damage of the positioners coils,
        // isolation and permanent magnets.
        // If an over load is detected the control-loop is disabled to protect the positioner.
        // WARNING
        // Electromagnetic driven positioners are not self-locking. Disabling the control-loop removes
        // any holding force from the positioner. Make sure not to damage any equipment
        // when the positioner unintentionally changes its position!

        // The present load level may be read in percent with the motor load property. This may be useful
        // to estimate the motor load while performing movements before the overload protection trips and
        // disables the control-loop. In case the motor load reaches a level close to 100 % the number of
        // movements per time, the movement acceleration and/or the mechanical load attached to the
        // positioner should be reduced.
        int32_t load;
        result = SA_CTL_GetProperty_i32(dHandle, channel, SA_CTL_PKEY_MOTOR_LOAD, &load, 0);
        exitOnError(result);
        printf("MCS2 short-term motor over load: %d %%\n", load);
    }
    // Disable the amplifier again.
    result = SA_CTL_SetProperty_i32(dHandle, channel, SA_CTL_PKEY_AMPLIFIER_ENABLED, SA_CTL_FALSE);
    exitOnError(result);
    printf("Done. Press return to exit.\n");
    getchar();

    // Before closing the program the connection to the device must be closed by calling "SA_CTL_Close".
    SA_CTL_Close(dHandle);
    printf("MCS2 close.\n");
    printf("*******************************************************\n");
    return 0;
}

File: drivers/SmarActMCS2/SDK-2.0.1/SDK/Python/examples/9_MCS2Example_MagneticDriver/MCS2Example_MagneticDriver.py
#!/usr/bin/python3
import sys
import smaract.ctl as ctl

# SmarAct MCS2 programming example: MagneticDriver
#
# This programming example demonstrates how to control
# electromagnetic driven positioners with the MCS2
#
# For a full command reference see the MCS2 Programmers Guide.

def assert_lib_compatibility():
    """
    Checks that the major version numbers of the Python API and the
    loaded shared library are the same to avoid errors due to 
    incompatibilities.
    Raises a RuntimeError if the major version numbers are different.
    """
    vapi = ctl.api_version
    vlib = [int(i) for i in ctl.GetFullVersionString().split('.')]
    if vapi[0] != vlib[0]:
        raise RuntimeError("Incompatible SmarActCTL python api and library version.")

def waitWhile(channel, mask):
    """ Poll channel state until masked flags are set"""
    while (True):
        state = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_STATE)
        if (state & mask) == 0: break

print("***************************************************************")
print("*  SmarAct MCS2 Programming Example (Electromagnetic Driver)  *")
print("***************************************************************")

# Read the version of the library
# Note: this is the only function that does not require the library to be initialized.
version = ctl.GetFullVersionString()
print("SmarActCTL library version: '{}'.".format(version))
assert_lib_compatibility()

# Find available MCS2 devices
try:
    buffer = ctl.FindDevices()
    if len(buffer) == 0:
        print("MCS2 no devices found.")
        sys.exit(1)
    locators = buffer.split("\n")
    for locator in locators:
        print("MCS2 available devices: {}".format(locator))
except:
    print("MCS2 failed to find devices. Exit.")
    input()
    sys.exit(1)

d_handle = None
try:
    # Open the first MCS2 device from the list
    d_handle = ctl.Open(locators[0])
    print("MCS2 opened {}.".format(locators[0]))

    channel = 0
    # Verify that there is an electromagnetic driven positioner available.
    type = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_TYPE) 
    state = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_STATE)
    if type != ctl.ChannelModuleType.MAGNETIC_DRIVER:
        print("MCS2 channel {} is not an electromagnetic driver channel, abort.".format(channel))
        ctl.Close(d_handle)
        input()
        sys.exit(1)
    elif (state & ctl.ChannelState.SENSOR_PRESENT) == 0:
        print("MCS2 no positioner connected to channel {}, abort.".format(channel))
        ctl.Close(d_handle)
        input()
        sys.exit(1)

    # The amplifier of electromagnetic driver channels is disabled at startup and must be explicitly enabled
    # before being able to perform closed-loop movements. This also implicitly starts the phasing sequence of
    # the positioner if it is not already phased.
    # The AMPLIFIER_ENABLED channel state bit reflects the state of the amplifier.
    if (state & ctl.ChannelState.AMPLIFIER_ENABLED) == 0:
        print("*******************************************************")
        print("-> Press return to enable the amplifier (and start the phasing sequence).")
        input()
        ctl.SetProperty_i32(d_handle, channel, ctl.Property.AMPLIFIER_ENABLED, ctl.TRUE)

        if (state & ctl.ChannelState.IS_PHASED) == 0:
            print("MCS2 phasing...")
            # Note that no external force or displacement must be applied to the positioner while the sequence is running.
            # The phasing takes some time to complete. During this time the ACTIVELY_MOVING
            # channel state bit is set.
            # The channel generates a PHASING_FINISHED event once the sequence has finished.
            # In this example we poll the channel state to determine the end of the phasing sequence instead of
            # listening to events.
            waitWhile(channel, ctl.ChannelState.ACTIVELY_MOVING)
            # The IS_PHASED channel state flag should then be checked to verify that the
            # phasing sequence finished successfully.
            state = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_STATE)
            if (state & ctl.ChannelState.IS_PHASED) == 0:
                print("MCS2 could not establish the phasing reference for channel {}, abort.".format(channel))
                ctl.Close(d_handle)
                input()
                sys.exit(1)
            else:
                print("MCS2 phase reference found.")

    # Perform referencing sequence if physical position is not known
    if (state & ctl.ChannelState.IS_REFERENCED) == 0:
        # Since the physical scale is not known yet no range limits can be used here.
        # Make sure to use a moderate move velocity for the referencing sequence.
        # Disable software range limits while referencing.
        ctl.SetProperty_i64(d_handle, channel, ctl.Property.RANGE_LIMIT_MIN, 0)
        ctl.SetProperty_i64(d_handle, channel, ctl.Property.RANGE_LIMIT_MAX, 0)
        ctl.SetProperty_i32(d_handle, channel, ctl.Property.REFERENCING_OPTIONS, 0)
        # The move velocity and acceleration properties also define the parameters for the referencing.
        ctl.SetProperty_i64(d_handle, channel, ctl.Property.MOVE_VELOCITY, 2000000000)
        ctl.SetProperty_i64(d_handle, channel, ctl.Property.MOVE_ACCELERATION, 100000000000)
        
        print("*******************************************************")
        print("-> Press return to start the referencing.")
        input()
        print("MCS2 referencing...")
        ctl.Reference(d_handle, channel)
        waitWhile(channel, ctl.ChannelState.REFERENCING)
        # Check if the referencing was successful.
        state = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_STATE)
        # The MOVEMENT_FAILED indicates a failed referencing.
        if (state & ctl.ChannelState.MOVEMENT_FAILED) != 0:
            # The channel error property may then be read to determine the reason of the error.
            error = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_ERROR)
            print("MCS2 referencing failed: {} (error: 0x{:04X}), abort.".format(ctl.GetResultInfo(error), error))
            ctl.Close(d_handle)
            input()
            sys.exit(1)
        else:
            print("MCS2 reference found.")
    # Note that electromagnetic driven positioners can reach very high velocities!
    # Moving into a physical endstop with high velocity can damage the positioner.
    # Therefore it is recommended to use the software range limits to define the active movement
    # range of the positioner. When commanding the positioner towards a limit the positioner is
    # decelerated to zero velocity in a way that it comes to a halt on the specified limit position.
    # Note: adjust this limits to match the movement range of your positioner!
    ctl.SetProperty_i64(d_handle, channel, ctl.Property.RANGE_LIMIT_MIN, -20000000000)
    ctl.SetProperty_i64(d_handle, channel, ctl.Property.RANGE_LIMIT_MAX, 20000000000)

    print("*******************************************************")
    print("-> Press return to start the movement.")
    input()
    # Set move mode to absolute movement.
    # Note: Open-loop movement is not available for electromagnetic driver channels.
    # (only the CL_ABSOLUTE and CL_RELATIVE move modes are valid)
    ctl.SetProperty_i32(d_handle, channel, ctl.Property.MOVE_MODE, ctl.MoveMode.CL_ABSOLUTE)
    # Set move velocity and acceleration.
    # Note that the velocity and acceleration control must be used for all movements. A velocity / acceleration
    # value of 0 (to disable the velocity / acceleration control) is invalid for magnetic driver channels.
    ctl.SetProperty_i64(d_handle, channel, ctl.Property.MOVE_VELOCITY, 10000000000)
    ctl.SetProperty_i64(d_handle, channel, ctl.Property.MOVE_ACCELERATION, 10000000000)

    positions = [2000000000,-2000000000,10000000000,-10000000000] # in pm
    for i in range(4):
        print("MCS2 move channel {} to position: {}...".format(channel, positions[i]))
        ctl.Move(d_handle, channel, positions[i])
        waitWhile(channel, ctl.ChannelState.ACTIVELY_MOVING)
        # Check if the movement was successful.
        state = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_STATE)
        # The MOVEMENT_FAILED channel state flag indicates a failed movement.
        if (state & ctl.ChannelState.MOVEMENT_FAILED) != 0:
            # The channel error property may then be read to determine the reason of the error.
            error = ctl.GetProperty_i32(d_handle, channel, ctl.Property.CHANNEL_ERROR)
            print("MCS2 movement failed: {} (error: 0x{:04X}).".format(ctl.GetResultInfo(error), error))
            break
            # Alternatively, the following channel state flags may be tested to determine the reason
            # of a failed movement:
            # - END_STOP_REACHED
            # - RANGE_LIMIT_REACHED
            # - FOLLOWING_LIMIT_REACHED
            # - POSITIONER_OVERLOAD
            # - POSITIONER_FAULT
            # See the MCS2 Programmers Guide for more information on the specific channel state flags.

        # The electromagnetic driver monitors the output current of each channel to detect an overload condition
        # of the positioner. This prevents thermal overheating and potential damage of the positioners coils,
        # isolation and permanent magnets.
        # If an over load is detected the control-loop is disabled to protect the positioner.
        # WARNING
        # Electromagnetic driven positioners are not self-locking. Disabling the control-loop removes
        # any holding force from the positioner. Make sure not to damage any equipment
        # when the positioner unintentionally changes its position!

        # The present load level may be read in percent with the motor load property. This may be useful
        # to estimate the motor load while performing movements before the overload protection trips and
        # disables the control-loop. In case the motor load reaches a level close to 100 % the number of
        # movements per time, the movement acceleration and/or the mechanical load attached to the
        # positioner should be reduced.
        load = ctl.GetProperty_i32(d_handle, channel, ctl.Property.MOTOR_LOAD)
        print("MCS2 short-term motor over load: {} %".format(load))

    # Disable the amplifier again.
    ctl.SetProperty_i32(d_handle, channel, ctl.Property.AMPLIFIER_ENABLED, ctl.FALSE)

except ctl.Error as e:
    # Passing an error code to "GetResultInfo" returns a human readable string
    # specifying the error.
    print("MCS2 {}: {}, error: {} (0x{:04X}) in line: {}."
          .format(e.func, ctl.GetResultInfo(e.code), ctl.ErrorCode(e.code).name, e.code, (sys.exc_info()[-1].tb_lineno)))

except Exception as ex:
    print("Unexpected error: {}, {} in line: {}".format(ex, type(ex), (sys.exc_info()[-1].tb_lineno)))
    raise

finally:
    # Before closing the program the connection to the device must be closed by calling "Close".
    if d_handle != None:
        ctl.Close(d_handle)
    print("MCS2 close.")
    print("*******************************************************")
    print("Done. Press return to exit.")
    input()
